pkg/resourcemanager/cosmosdb/sqldatabase/cosmosdbsql.go

```diff
@@ -0,0 +1,275 @@
+		instance.Spec.ResourceGroup,
+		instance.Spec.CosmosDBAccount,
+		instance.ObjectMeta.Name)
+	if err != nil {
+		azerr := errhelp.NewAzureError(err)
+		if azerr.Type != errhelp.ResourceNotFound && azerr.Type != errhelp.NotFoundErrorCode {
+			instance.Status.Message = fmt.Sprintf("failed GET-ing CosmosDBSQLDatabaseThroughput: %s", err.Error())
+			return errhelp.IsErrorFatal(err, notFoundErrorCodes, nil)
+		}
+	}
+
+	// Transform our spec to Azure shape
+	transformed := m.transformToCosmosDBSQLDatabase(instance)
+
+	// Check if we're in the goal state already
+	resourceMatchesAzure := DoesResourceMatchAzure(transformed, sqlDatabase, sqlDatabaseThroughputSettings)
+	if resourceMatchesAzure {
+		instance.Status.SetProvisioned(resourcemanager.SuccessMsg)
+		instance.Status.ResourceId = *sqlDatabase.ID
+		return true, nil
+	}
+
+	// Drive to goal state
+	instance.Status.SetProvisioning("")
+	future, err := cosmosDBSQLDatabaseClient.CreateUpdateSQLDatabase(
+		ctx,
+		instance.Spec.ResourceGroup,
+		instance.Spec.CosmosDBAccount,
+		instance.ObjectMeta.Name,
+		transformed)
+	if err != nil {
+		instance.Status.Message = err.Error()
+		allowedErrors := []string{
+			errhelp.AsyncOpIncompleteError,
+			errhelp.AlreadyExists,
+			errhelp.FailoverGroupBusy,
+		}
+		unrecoverableErrors := []string{
+			errhelp.InvalidFailoverGroupRegion,
+		}
+		return errhelp.IsErrorFatal(err, append(allowedErrors, notFoundErrorCodes...), unrecoverableErrors)
+	}
+	instance.Status.SetProvisioning("Resource request successfully submitted to Azure")
+	instance.Status.PollingURL = future.PollingURL()
+
+	// Need to poll the polling URL, so not done yet!
+	return false, nil
+}
+
+func (m *AzureCosmosDBSQLDatabaseManager) Delete(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
+	options := &resourcemanager.Options{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	instance, err := m.convert(obj)
+	if err != nil {
+		return false, err
+	}
+
+	cosmosDBSQLDatabaseClient, err := cosmosdb.GetCosmosDBSQLDatabaseClient(m.Creds)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to create cosmos DB SQL database client")
+	}
+
+	_, err = cosmosDBSQLDatabaseClient.DeleteSQLDatabase(ctx, instance.Spec.ResourceGroup, instance.Spec.CosmosDBAccount, instance.ObjectMeta.Name)
+	if err != nil {
+		instance.Status.Message = err.Error()
+		azerr := errhelp.NewAzureError(err)
+
+		// these errors are expected
+		ignore := []string{
+			errhelp.AsyncOpIncompleteError,
+		}
+
+		// this means the thing doesn't exist
+		finished := []string{
+			errhelp.ResourceNotFound,
+			errhelp.ResourceGroupNotFoundErrorCode,
+		}
+
+		if helpers.ContainsString(ignore, azerr.Type) {
+			return true, nil
+		}
+
+		if helpers.ContainsString(finished, azerr.Type) {
+			return false, nil
+		}
+		instance.Status.Message = fmt.Sprintf("delete failed with: %s", err.Error())
+		return false, err
+	}
+
+	// Not done yet, keep trying
+	return false, nil
```

Comment:
Shouldn't this be `true, nil`? I think the sense of this return is "does the object still exist?", and returning false here causes AsyncReconciler to remove the finalizer.